File: CoilCombination/SVD_cc.m
function Image_cc = SVD_cc(Image_mc)
% SVD based Coil Combination for Hyperpolarized MRI data
% Reference:
%   Hall, Emma L., et al. "Methodology for improved detection of low 
%   concentration metabolites in MRS: optimised combination of signals from 
%   multi-element coil arrays." Neuroimage 86 (2014): 35-42.
%
% INPUTS:
%   Image_mc - multi-channel image data. dim:[Nx, Ny, Nz, Ncoil, 1,
%       N_freq, Ndyn]
%       
% OUTPUTS:
%   Image_cc - coil combination Image.
%
% (c) 2017-2018 The Regents of the University of California
% All Rights Reserved.
%
% Written by :Xucheng Zhu
%   

[Nx, Ny, Nz, Ncoil, Nmap, N_freq, Ndyn] = size(Image_mc);
Image_mc = prewhitening_cc(Image_mc);
Image_cc = zeros([Nx, Ny, Nz, 1, Nmap, N_freq, Ndyn]);

for i = 1:Nx
    for j = 1:Ny
        for k = 1:Nz
            for t = 1:Ndyn
                tmp_spectrum = Image_mc(i,j,k,:,1,:,t);
                tmp_spectrum = squeeze(tmp_spectrum);
                [U,S,V] = svd(tmp_spectrum,'econ');
                S = wthresh(S,1);
                Image_cc(i,j,k,1,1,:,t) = sum(U*S*V',2);
            end
        end
    end
end

File: reconstruction/CoilCombination/Python/SVD_cc.py
import numpy as np
from scipy.linalg import svd
from skimage.restoration import denoise_wavelet

def SVD_cc(image_mc):
    
    #  SVD based Coil Combination for Hyperpolarized MRI data
    #  Reference:
    #    Hall, Emma L., et al. "Methodology for improved detection of low 
    #    concentration metabolites in MRS: optimised combination of signals from 
    #    multi-element coil arrays." Neuroimage 86 (2014): 35-42.

    #  INPUTS:
    #    Image_mc - multi-channel image data. dim:[Nx, Ny, Nz, Ncoil, N_freq, Ndyn]
        
    #  OUTPUTS:
    #    Image_cc - coil combination Image.

    # (c) 2017-2018 The Regents of the University of California
    # All Rights Reserved.

    # Written by :Parimal Joshi 

    Nx, Ny, Nz, Ncoil, N_freq, Ndyn = image_mc.shape
    image_cc = np.zeros((Nx, Ny, Nz, N_freq, Ndyn), dtype=image_mc.dtype)

    for i in range(Nx):
        for j in range(Ny):
            for k in range(Nz):
                for t in range(Ndyn):
                    tmp_spectrum = image_mc[i, j, k, :, :, t]  # [Ncoil, N_freq]
                    U, s, Vh = svd(tmp_spectrum, full_matrices=False)
                    thresh = s * (np.abs(s) > 1)
                    S = np.diag(thresh)
                    combined = np.sum(U @ S @ Vh, axis=0)  # [N_freq]
                    image_cc[i, j, k, :, t] = combined

    return image_cc
